### PR TITLE
Disallow dropping file outside of the Dropzone by default (#87)

### PR DESCRIPTION
**Pull request by mwolson**
* New option `allowGlobalDrops` - if true - will allow dropped items to potentially take over the browser window
* If false, the new default behavior, it will instead install document event handlers to prevent this from happening

### CODE REVIEW

**Comment by okonet** [comment]
Thanks for your PR!

> New option allowGlobalDrops - if true - will allow dropped items to potentially take over the browser window

Not sure how introducing the new prop will solve the #87. Why do you think this can be useful?

**Comment by mwolson** [comment]
In this PR, there are some addEventListener calls for drop and dragover. The new prop controls whether those get installed. When installed (this happens by default, false case) these hooks make the entire browser window an allowable drag-and-drop target, but don't do anything with the drop. This overrides the default browser behavior of navigating to the dropped item.

**Comment by okonet** [comment]
@mwolson great work on this PR! My question is if we even need a prop o control that behavior? I can't imagine when the browser default behavior would be desirable in case there is dropzone installed. Do you?

Second thing: can you add a test where multiple dropzones are being used on the same page and check if it still passes? Did you actually test in the real browser as well? I know this is very cumbersome at the moment.

**Comment by mwolson** [comment]
I'd be OK with removing the prop if you want, since the default browser behavior is definitely not desirable.

I've tested it in a real browser (Firefox and Chrome, not yet IE/Edge) with a single dropzone. I'll add a unit test with multiple dropzones shortly.

**Comment by mwolson** [comment]
One possibility for why we might need a prop in the future: if the user has 1000+ very granular dropzones for some reason, they may want to install the global document handling only once. But I don't know how probable it would be for someone to do that.

**Comment by mwolson** [comment]
After testing, the document drop handler is capturing events meant for Dropzone components, will need to handle targets of components that have gone away during re-render, and parent Dropzone components need special handling to reset their state when child Dropzone components get a drop. Going to work further to address these before proceeding with a merge.

**Comment by mwolson** [comment]
Handled child/parent drops in 4b2d86b on my other PR.

**Comment by mwolson** [comment]
Rebased this on top of my other branch since there's starting to be unavoidable overlap. Finished addressing all of the remaining issues with both PR's. Removed WIP status - ready for another look.

File: README.md [hunk 1]
```diff
@@ -92,6 +92,7 @@ You have a third option : providing a function that returns the component's chil
 Features
 ========
 
+- `allowGlobalDrops` `[Boolean | **false**]` — Dropping files outside of the `<Dropzone>` is allowed. This can cause the current browser window to be taken over by the dropped file.
```
Comment (okonet) [inline README.md]: Should this be renamed to something like `preventDropOnDocument` with the default value of `true`? I think right the purpose of it is not clear enough.

File: src/index.js [hunk 2]
```diff
@@ -19,30 +24,52 @@ class Dropzone extends React.Component {
...
     };
   }
 
   componentDidMount() {
-    this.enterCounter = 0;
+    const { allowGlobalDrops } = this.props;
+    this.dragTargets = [];
+
+    if (!allowGlobalDrops) {
+      document.addEventListener('dragover', Dropzone.onDocumentDragOver, false);
```
Comment (okonet) [inline src/index.js]: Why do we need the `dragover` handler?

File: src/index.js [hunk 3]
```diff
@@ -102,39 +132,47 @@ class Dropzone extends React.Component {
...
+    this.dragTargets = [];
     this.isFileDialogActive = false;
 
-    fileList.forEach((file) => {
-      if (!disablePreview) {
-        file.preview = window.URL.createObjectURL(file); // eslint-disable-line no-param-reassign
-      }
+    // Don't fire callbacks if a child Dropzone component has already handled this event
+    // eslint-disable-next-line no-underscore-dangle
+    if (!e.__dropzoneChildHandledThis) {
```
Comment (okonet) [inline src/index.js]: Hmm, I don't like how this looks TBH. Why not use `event.stopPropagation` instead?

File: src/index.js [hunk 4]
```diff
@@ -102,39 +132,47 @@ class Dropzone extends React.Component {
...
+      e.persist();
```
Comment (okonet) [inline src/index.js]: Can you point me at what this does?

File: src/index.spec.js [hunk 5]
```diff
@@ -37,6 +49,53 @@ describe('Dropzone', () => {
       expect(dropzone.html()).toMatchSnapshot();
     });
 
+    function collectEventListenerCalls(calls) {
```
Comment (okonet) [inline src/index.spec.js]: Please add a JSDoc comment for this function. I hardly understand what it does and why it's needed.

File: src/index.spec.js [hunk 6]
```diff
@@ -37,6 +49,53 @@ describe('Dropzone', () => {
...
+      return calls.reduce((acc, [eventName, ...rest]) => {
+        acc[eventName] = rest; // eslint-disable-line no-param-reassign
+        return acc;
+      }, {});
+    }
+
+    it('prevents stray drops from taking over the browser window by default', () => {
+      const dropzone = mount(<Dropzone><p>Content</p></Dropzone>);
+      expect(dropzone.html()).toMatchSnapshot();
+      expect(document.addEventListener.callCount).toEqual(2);
```
Comment (okonet) [inline src/index.spec.js]: It's probably better to test in a separate `it` block with an appropriate description. Either way, shouldn't it test explicitly what handlers were attached?

File: src/index.spec.js [hunk 7]
```diff
@@ -37,6 +49,53 @@ describe('Dropzone', () => {
...
+      const addCalls = collectEventListenerCalls(document.addEventListener.args);
```
Comment (okonet) [inline src/index.spec.js]: ?

File: src/index.spec.js [hunk 8]
```diff
@@ -37,6 +49,53 @@ describe('Dropzone', () => {
...
+
+      const event = { preventDefault: spy() };
```
Comment (okonet) [inline src/index.spec.js]: I wondering if this test should be that detailed. I think it will be enough to test that we've called the callback function for appropriate edge cases?

File: src/index.spec.js [hunk 9]
```diff
@@ -317,23 +376,59 @@ describe('Dropzone', () => {
       dropzone.simulate('dragEnter', { dataTransfer: { files } });
       expect(dropzone.text()).toEqual('Rejected');
     });
+
+    it('should reset the dragActive/dragReject state when leaving after a child goes away', () => {
```
Comment (okonet) [inline src/index.spec.js]: Wasn't it the separate issue?

File: src/index.spec.js [hunk 10]
```diff
@@ -498,6 +593,144 @@ describe('Dropzone', () => {
       expect(dropRejectedSpy.callCount).toEqual(0);
     });
 
+    it('accepts files on the innermost Dropzone when we drop into it', () => {
```
Comment (okonet) [inline src/index.spec.js]: Not sure what this is testing. What is `innermost` Dropzone?

File: README.md (hunk 1 above)
Comment (mwolson) [inline README.md]: `preventDropOnDocument` seems fine

File: src/index.js (hunk 2 above)
Comment (mwolson) [inline src/index.js]: If `dragover` is not provided, dropped items will still take over the browser window, because the browser doesn't consider a div with only a `drop` event (and no `dragover`) to be a valid `drop` target.

File: src/index.js (hunk 3 above)
Comment (mwolson) [inline src/index.js]: The problem is that the parent needs to know whether some child has already fired an onDrop() callback. If a child has fired, reset the dragActive/dragReject/dragTargets state and don't fire onDrop(). If a child has not fired, reset the state and do fire onDrop().

If we used `event.stopPropagation()`, the parent would never receive the event, so its state would remain active, and the render of the parent would still be styled with `dragActive=true` after a drop on the child.

The order is like:
1: Document `drop` handler (if installed via preventGlobalDrops=true)
2: Child `drop` handler
3: Parent `drop` handler

File: src/index.js (hunk 4 above)
Comment (mwolson) [inline src/index.js]: Normally event objects that React passes to event handlers are structurally reused, like:

```js
delete e.target;
```

And then put in a pool. If another event comes in, they'll get removed from the pool and sent to another handler.

`e.persist()` will keep that event out of the pool, so we get our own copy of it, and can mutate the event by adding the `__dropzoneChildHandledThis` property. When I tried to do this without an `e.persist()` call, React would throw a warning to the browser console.

Further reading: https://facebook.github.io/react/docs/events.html

File: src/index.spec.js (hunk 5 above)
Comment (mwolson) [inline src/index.spec.js]: OK I'll add JSDoc

File: src/index.js (hunk 3 above)
Comment (okonet) [inline src/index.js]: Wait a minute, are you talking about nested Dropzones? Why would we support that? What's the use case?

File: src/index.spec.js (hunk 6 above)
Comment (mwolson) [inline src/index.spec.js]: I'll split this into its own `describe` block I think, with multiple `it` steps.

File: src/index.spec.js (hunk 8 above)
Comment (mwolson) [inline src/index.spec.js]: The problem is that there's no way to fire an event on `document`. So if we didn't manually exercise it here, we wouldn't get test coverage. And it's useful to make sure that we don't call `e.preventDefault()` in the case where the drag event is meant for one of our dropzones - because if we did call `e.preventDefault()` in that case, the dropzone would not get the event.

```
1: document onDrop() <-- stops propagation -->
2: dropzone instance onDrop() <-- doesn't get event now -->
```

File: src/index.spec.js (hunk 10 above)
Comment (mwolson) [inline src/index.spec.js]: The innermost dropzone is the child Dropzone component. The outermost dropzone is the parent Dropzone component.

Comment (okonet) [inline src/index.spec.js]: Yeah, as I already asked ↑ do we really need to support that case? It seems very edgy to me and the complexity it brings to the project doesn't seem worth it. What's the use case for it?

File: src/index.js (hunk 3 above)
Comment (mwolson) [inline src/index.js]: Yeah it's nested Dropzones. An example would be: an instant messenger UI where dragging over an entire messaging window will show an overlay modal, and then you may potentially be able to drop into one or more large icons to indicate whether to send the attachment immediately or compose a message to go with it, with a default behavior of "compose" if the user drops into the modal itself and misses the icons.

Comment (okonet) [inline src/index.js]: I don't believe the complexity that we need to bring in to support such a edge case is worth it TBH. Let's keep things simple as long as possible. When I was asking about multiple dropzones on the page I was referring to a use case with a list of dropzones for instance. Are you okay with that?

File: src/index.js (hunk 4 above)
Comment (okonet) [inline src/index.js]: Interesting. I'll read but see my ↑ comment.

File: src/index.js (hunk 3 above)
Comment (mwolson) [inline src/index.js]: OK. I don't immediately need nested dropzone support for my own use case yet, but might in the future. I think it might be surprising for someone who tries nesting to find that the parent still receives a drop event, but it's hard to predict whether anyone will actually want to do that. I'll go ahead and remove it from the 2 PR's and stash the changes somewhere; maybe it could be a separate PR by itself later if needed.

I tested the sibling case manually, and it works fine. A single user drop gets fired like:

```
event chain 1: sibling 1 document drop: <-- this isn't me, e.preventDefault() -->
event chain 2: sibling 2 document drop: <-- don't call e.preventDefault() -->
event chain 2: sibling 2 dropzone instance: <-- receives event -->
```

Comment (okonet) [inline src/index.js]: Awesome!

> I think it might be surprising for someone who tries nesting to find that the parent still receives a drop event

TBH I would be surprised if it wouldn't receive and fire the event :)

**Comment by mwolson** [comment]
Finished one round of changes, still need to split out parent/child nesting, will do later today.

**Comment by mwolson** [comment]
Finished rebasing this on top of the other updated PR, so there should be no more handling of nested components (except for the new tests which document the current behavior).

**Comment by okonet** [comment]
Can you please rebase on top of master?

**Comment by mwolson** [comment]
Sure thing, finished the rebase onto master

**Comment by okonet** [comment]
@mwolson just wondering if you also tries this example:

```js
function dragover_handler(ev) {
  console.log("dragOver");
  // Prevent default select and drag behavior
  ev.preventDefault();
}
```

from https://developer.mozilla.org/en-US/docs/Web/API/HTML_Drag_and_Drop_API/File_drag_and_drop

It says it should prevent default browser behavior.

**Comment by mwolson** [comment]
@okonet that example prevents default browser dragover behavior, but only for drags on top of that div. It doesn't do anything for drags on the rest of the document, so we still need the `document` handlers.

**Comment by okonet** [comment]
So even if this handler is on `document`? I'm just wondering if we can go away with just one handler on the document.

**Comment by tt-mikeolson** [comment]
Both handlers are needed - take either away and the feature stops working.

**Comment by mwolson** [comment]
Both handlers are needed - take either away and the feature stops working.

**Comment by okonet** [comment]
Gotcha! Thanks!
